TooltipDecorator/useTooltip.js

```diff
@@ -0,0 +1,266 @@
+
+		newLayoutInfo.tooltipDirection = adjustDirection(newLayoutInfo.tooltipDirection, overflow, rtl);
+		newLayoutInfo.arrowAnchor = adjustAnchor(newLayoutInfo.arrowAnchor, newLayoutInfo.tooltipDirection, overflow, rtl);
+		newLayoutInfo.position = getPosition(clientNode, newLayoutInfo.tooltipDirection);
+		newLayoutInfo.labelOffset = newLayoutInfo.arrowAnchor === 'center' ? getLabelOffset(tooltipNode, newLayoutInfo.tooltipDirection, newLayoutInfo.position, overflow) : null;
+
+		if (
+			(newLayoutInfo.position.top !== layoutInfo.position.top) ||
+			(newLayoutInfo.position.left !== layoutInfo.position.left) ||
+			(newLayoutInfo.labelOffset !== layoutInfo.labelOffset) ||
+			(newLayoutInfo.arrowAnchor !== layoutInfo.arrowAnchor)
+		) {
+			setLayoutInfo(newLayoutInfo);
+		}
+	}, [tooltipText, rtl, tooltipPosition, tooltipType, screenEdgeKeepout, layoutInfo, tooltipRef, clientRef]);
+
+	useLayoutEffect(() => {
+		mutableRef.current.showTooltipJob = new Job(() => {
+			if (!showing) {
+				setShowing(true);
+			}
+		});
+	}, [showing]);
+
+	useLayoutEffect(() => {
+		mutableRef.current.setTooltipLayoutJob = new Job(() => {
+			setTooltipLayout();
+		});
+	}, [setTooltipLayout]);
+
+	const getTooltipRef = useCallback((node) => {
+		tooltipRef.current = node;
+		if (node) {
+			setTooltipLayout();
+		}
+	}, [setTooltipLayout]);
+
+	const renderTooltip = useCallback(() => {
+		const {top, left} = layoutInfo.position;
+		const tooltipStyle = {
+			// Moving the position to CSS variables where there are additional offset calculations
+			'--tooltip-position-top': tooltipRelative ? null : ri.unit(top, 'rem'),
+			'--tooltip-position-left': tooltipRelative ? null : ri.unit(left, 'rem')
+		};
+
+		/**
+		 * Conditionally creates the FloatingLayer and Tooltip based on the presence of
+		 * `tooltipText` and returns a property bag to pass onto the Wrapped component
+		 *
+		 * @returns {Object} Prop object
+		 * @private
+		 */
+		const renderedTooltip = (
+			<Tooltip
+				aria-hidden
+				labelOffset={layoutInfo.labelOffset}
+				{...tooltipProps}
+				arrowAnchor={layoutInfo.arrowAnchor}
+				direction={layoutInfo.tooltipDirection}
+				marquee={tooltipMarquee}
+				relative={tooltipRelative}
+				style={tooltipStyle}
+				tooltipRef={getTooltipRef}
+				type={tooltipType}
+				width={tooltipWidth}
+			>
+				{tooltipText}
+			</Tooltip>
+		);
+
+		if (!tooltipRelative) {
+			return (
+				<FloatingLayerBase open={showing} noAutoDismiss onDismiss={hideTooltip} scrimType="none" key="tooltipFloatingLayer">
+					{renderedTooltip}
+				</FloatingLayerBase>
+			);
+		} else if (showing) {
+			return renderedTooltip;
+		} else {
+			return null;
+		}
+	}, [getTooltipRef, hideTooltip, layoutInfo.arrowAnchor, layoutInfo.labelOffset, layoutInfo.position, layoutInfo.tooltipDirection, showing, tooltipMarquee, tooltipProps, tooltipRelative, tooltipText, tooltipType, tooltipWidth]);
+
+	return {
+		tooltipChildren: tooltipText ? renderTooltip() : null,
```

Comment:
Would you explain why you didn't add `tooltipDestinationProp` to useTooltip?

Reply:
As a HOC, TooltipDecorator must get the name of a prop passed to a wrapped component since HOC does not know a wrapped component's props.
But, when a parent functional component or a parent(?) hook uses `useTooltip`, a parent knows the name of the right prop and handles the prop directly. Just `tooltipChildren` is required.